api/src/test/java/org/openmrs/api/EncounterServiceTest.java

```diff
@@ -98,7 +97,235 @@
+		ObsService os = Context.getObsService();
+
+		Obs o1 = createObs(cs.getConcept(3),"original obs value text");
+		Obs o2 = createObs(cs.getConcept(3),"second obs value text");
+		Obs oParent =createObs(cs.getConcept(23), null); //in the concept set table as a set
+
+		//add o1 and o2 to the parent obs
+		oParent.addGroupMember(o2);
+		oParent.addGroupMember(o1);
+
+		Encounter enc = buildEncounter();
+		enc.addObs(oParent);
+
+		enc = es.saveEncounter(enc);
+
+		//Create a new obs o4 and add it to parent
+		oParent = os.getObs(oParent.getObsId()); // Refresh the oParent
+		oParent.removeGroupMember(os.getObs(o2.getObsId())); // remove o2 from oParent
+
+		enc = es.saveEncounter(enc);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		assertEquals(1,obsAtTopLevelUpdated.size());
+
+		assertEquals(1, obsAtTopLevelUpdated.iterator().next().getGroupMembers(true).size()); // should be o1
+
+		oParent = os.getObs(oParent.getObsId());
+		assertTrue(oParent.getGroupMembers(true).contains(os.getObs(o1.getObsId())));
+
+		assertNull(os.getObs(o2.getObsId()).getObsGroup());
+	}
+
+
+	@Test
+	@Verifies(value = "should save encounter correctly when a level 1 child obs is updated and there are multiple top level obs without recreating an entire obs hierarchy", method = "saveEncounter(Encounter)")
+	public void saveEncounter_shouldSaveEncounterWhenLevel1ObsIsUpdatedAndThereAreMultipleTopLevelObs(){
+		ConceptService cs = Context.getConceptService();
+		EncounterService es = Context.getEncounterService();
+		ObsService os = Context.getObsService();
+
+		Obs o1 = createObs(cs.getConcept(3),"original obs value text");
+		Obs o2 = createObs(cs.getConcept(3),"second obs value text");
+		Obs oParent = createObs(cs.getConcept(23), null); //in the concept set table as a set
+		Obs o3 = createObs(cs.getConcept(3), "third obs value text");
+
+		//add o and o2 to the parent obs
+		oParent.addGroupMember(o2);
+		oParent.addGroupMember(o1);
+
+		Encounter enc = buildEncounter();
+		enc.addObs(oParent);
+		enc.addObs(o3);
+
+		enc = es.saveEncounter(enc);
+
+		Obs o2Updated = os.getObs(o2.getObsId());
+		o2Updated.setValueText("Updated second obs value text");
+
+		enc = es.saveEncounter(enc);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		assertEquals(2,obsAtTopLevelUpdated.size()); //oParent and o3
+
+		oParent = os.getObs(oParent.getObsId());
+		assertEquals(3,oParent.getGroupMembers(true).size());
+
+		assertTrue(os.getObs(o2.getObsId()).getVoided());
+		o2Updated = getNewVersionOfEditedObs(oParent, o2Updated);
+		assertFalse(o2Updated.getVoided());
+	}
+
+	private Obs createObs(Concept concept, String value){
+		Obs o = new Obs();
+		o.setConcept(concept);
+		o.setDateCreated(new Date());
+		o.setCreator(Context.getAuthenticatedUser());
+		o.setLocation(new Location(1));
+		o.setObsDatetime(new Date());
+		o.setPerson(new Patient(3));
+
+		if(value!=null){
+			o.setValueText(value);
+		}
+
+		return o;
+	}
+
```

Comment:
I really don't think you needed to have tests for top level vs nested child, I think you should have tests for removing, adding or editing a child regardless of the level